torchao/prototype/mx_formats/fp_formats.py

```diff
@@ -0,0 +1,550 @@
+            # 2. float8_e4m3fn only sets {s}.{1111}.{111} for nan
+            if all(b == "1" for b in e_enc + m_enc):
+                return None, None, None, float("nan")
+
+    exponent = exp_encoding_to_exp(e_enc, dtype)
+
+    is_zero = all(b == "0" for b in e_enc + m_enc)
+    is_denormal = (not is_zero) and all(b == "0" for b in e_enc)
+    is_normal = not is_zero and not is_denormal
+
+    if is_zero:
+        return sign, exponent, 0.0, None
+
+    mantissa = 1.0 if is_normal else 0.0
+    cur_pow_2 = -1
+    for m_bit in m_enc:
+        mantissa += int(m_bit) * pow(2, cur_pow_2)
+        cur_pow_2 -= 1
+    return sign, exponent, mantissa, None
+
+
+def sem_vals_to_f32(s_i, e_i, m_f, special_value):
+    """
+    Input: integer sign, integer exponent, float32 mantissa, special value
+    Output: float32 value
+    """
+    if special_value is not None:
+        return special_value
+    f = s_i * pow(2, e_i) * m_f
+    return f
+
+
+def sem_vals_to_formula(s_i, e_i, m_f, special_value):
+    """
+    Input: integer sign, integer exponent, float32 mantissa, special value
+    Output: formula to get the float32 value
+    """
+    if special_value is not None:
+        return special_value
+    return f"{s_i} * 2^{e_i} * {m_f}"
+
+
+def assert_same(fp1, fp2):
+    if math.isnan(fp1):
+        assert math.isnan(fp2)
+    elif math.isinf(fp1):
+        if fp1 > 0:
+            assert math.isinf(fp2) and fp2 > 0
+        else:
+            assert math.isinf(fp2) and fp2 < 0
+    else:
+        assert (abs(fp2 - fp1) / (fp1 + 1e-20)) - 1 < 1e-12, f"{fp2} != {fp1}"
+
+
+def run(dtype):
+    print("dtype", dtype)
+
+    headers = ["orig_val", "formula", "s_enc", "e_enc", "m_enc", "note"]
+    results = []
+
+    if dtype == DTYPE_FP4:
+        results = float4_e2m1_interesting_values
+    elif dtype == DTYPE_FP6_E3M2:
+        results = float6_e3m2_interesting_values
+    elif dtype == DTYPE_FP6_E2M3:
+        results = float6_e2m3_interesting_values
+    else:
+        interesting_values = dtype_to_interesting_values[dtype]
+        for row in interesting_values:
+            fp_ref, s_enc_ref, e_enc_ref, m_enc_ref, notes = row
+
+            # test that things still work
+            _assert_equals(fp_ref, s_enc_ref, e_enc_ref, m_enc_ref, dtype)
+
+            # create the formula
+            s_i, e_i, m_f, special_value = sem_bits_to_sem_vals(
+                s_enc_ref, e_enc_ref, m_enc_ref, dtype
+            )
+            formula = sem_vals_to_formula(s_i, e_i, m_f, special_value)
+
+            # create the table row
+            results.append(
+                [
+                    fp_ref,
+                    formula,
+                    s_enc_ref,
+                    e_enc_ref,
+                    m_enc_ref,
+                    notes,
+                ]
+            )
+
+    print(tabulate.tabulate(results, headers=headers))
+    print("\n")
+
+
+if __name__ == "__main__":
```

Comment:
wdyy about renaming this file to have spec in the name? I quite like and we can recommend people to cross reference the text spec with your code in the main README